### PR TITLE
Remove SDS Timeout

### PR DESCRIPTION
**Pull request by adiprerepa**
[X ] Networking

Sorry, Looks like I messed up git. Hopefully this gets merged quick now.

cc @JimmyCYJ @howardjohn 

again, sorry, hope this is everything
https://github.com/istio/istio/issues/25339

### CODE REVIEW

**Comment by adiprerepa** [comment]
@myidpt its the same as last time, ptal

File: pilot/pkg/security/model/authentication.go [hunk 1]
```diff
@@ -150,6 +154,13 @@ func ConstructSdsSecretConfig(name string) *tls.SdsSecretConfig {
 		return nil
 	}
 
+	var fetchTimeout *durationpb.Duration
+	if name == SDSDefaultResourceName || name == SDSRootResourceName {
+		fetchTimeout = ptypes.DurationProto(time.Second * 0)
+	} else {
+		fetchTimeout = features.InitialFetchTimeout
+	}
+
```
Comment (williamaronli) [inline pilot/pkg/security/model/authentication.go:163]: I am wondering why you need to add this code logic here. As from the code if name == SDSDefaultResourceName or SDSRootResourceName the ConstructSdsSecretConfig function  will directly return the defaultSDSConfig and rootSDSConfig. So you only need to change the var field InitialFetchTimeout  of defaultSDSConfig and SDSRootResourceName to ptypes.DurationProto(time.Second * 0)

**Comment by williamaronli** [comment]
@adiprerepa I will take a look. just ping me I will review your code

**Comment by williamaronli** [comment]
@adiprerepa , please add a release note here like example code here: https://github.com/istio/istio/blob/f97b9da3c7354b2d763f4b3b2894ae33bdd81bc4/releasenotes/notes/ecc-csr.yaml#L6

**Comment by williamaronli** [comment]
@adiprerepa could you address my comment?

**Comment by adiprerepa** [comment]
@williamaronli sorry, addressing.

File: pilot/pkg/security/model/authentication.go (hunk 1 above)
Comment (adiprerepa) [reply pilot/pkg/security/model/authentication.go:163]: @howardjohn indicated this was necessary, but I think you are right. Will change.

Comment (adiprerepa) [reply pilot/pkg/security/model/authentication.go]: oh, i see what happened. When I rebased this PR the logic to return rootSDSConfig or defaultSDSConfig were added, making my if statement redundant. OK, i removed it.

**Comment by adiprerepa** [comment]
@williamaronli added a release-notes file.

**Comment by adiprerepa** [comment]
anyone know why TestGolden is failing? @williamaronli @howardjohn

**Comment by howardjohn** [comment]
Suggest this diff:

```diff
diff --git a/pkg/bootstrap/instance_test.go b/pkg/bootstrap/instance_test.go
index 576c74b968..b2e7642945 100644
--- a/pkg/bootstrap/instance_test.go
+++ b/pkg/bootstrap/instance_test.go
@@ -398,9 +398,8 @@ func TestGolden(t *testing.T) {
 
 			checkOpencensusConfig(t, realM, goldenM)
 
-			if !reflect.DeepEqual(realM, goldenM) {
-				s, _ := diff.PrettyDiff(goldenM, realM)
-				t.Logf("difference: %s", s)
+			if diff := cmp.Diff(goldenM, realM, protocmp.Transform()); diff != "" {
+				t.Logf("difference: %s", diff)
 				t.Fatalf("\n got: %s\nwant: %s", prettyPrint(read), prettyPrint(jgolden))
 			}
 
diff --git a/pkg/bootstrap/testdata/all_golden.json b/pkg/bootstrap/testdata/all_golden.json
index 943ff228d5..cf5d3f0883 100644
--- a/pkg/bootstrap/testdata/all_golden.json
+++ b/pkg/bootstrap/testdata/all_golden.json
@@ -430,7 +430,7 @@
         "name": "envoy_metrics_service",
         "type": "STRICT_DNS",
       
-        "transport_socket": {"name":"envoy.transport_sockets.tls","typed_config":{"@type":"type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext","common_tls_context":{"alpn_protocols":["h2"],"combined_validation_context":{"default_validation_context":{},"validation_context_sds_secret_config":{"name":"file-root:/etc/istio/ms/ca.pem","sds_config":{"api_config_source":{"api_type":"GRPC","grpc_services":[{"envoy_grpc":{"cluster_name":"sds-grpc"}}],"transport_api_version":"V3"},"resource_api_version":"V3"}}},"tls_certificate_sds_secret_configs":[{"name":"file-cert:/etc/istio/ms/client.pem~/etc/istio/ms/key.pem","sds_config":{"api_config_source":{"api_type":"GRPC","grpc_services":[{"envoy_grpc":{"cluster_name":"sds-grpc"}}],"transport_api_version":"V3"},"resource_api_version":"V3"}}]},"sni":"envoy_metrics_service"}},
+        "transport_socket": {"name":"envoy.transport_sockets.tls","typed_config":{"@type":"type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext","common_tls_context":{"alpn_protocols":["h2"],"combined_validation_context":{"default_validation_context":{},"validation_context_sds_secret_config":{"name":"file-root:/etc/istio/ms/ca.pem","sds_config":{"api_config_source":{"api_type":"GRPC","grpc_services":[{"envoy_grpc":{"cluster_name":"sds-grpc"}}],"transport_api_version":"V3"},"initial_fetch_timeout":"0s","resource_api_version":"V3"}}},"tls_certificate_sds_secret_configs":[{"name":"file-cert:/etc/istio/ms/client.pem~/etc/istio/ms/key.pem","sds_config":{"api_config_source":{"api_type":"GRPC","grpc_services":[{"envoy_grpc":{"cluster_name":"sds-grpc"}}],"transport_api_version":"V3"},"initial_fetch_timeout":"0s","resource_api_version":"V3"}}]},"sni":"envoy_metrics_service"}},
       
       
         "respect_dns_ttl": true,
diff --git a/pkg/bootstrap/testdata/tracing_tls_golden.json b/pkg/bootstrap/testdata/tracing_tls_golden.json
index 218770230c..b187c2ec51 100644
--- a/pkg/bootstrap/testdata/tracing_tls_golden.json
+++ b/pkg/bootstrap/testdata/tracing_tls_golden.json
@@ -370,7 +370,7 @@
       {
         "name": "zipkin",
         
-        "transport_socket": {"name":"envoy.transport_sockets.tls","typed_config":{"@type":"type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext","common_tls_context":{"combined_validation_context":{"default_validation_context":{},"validation_context_sds_secret_config":{"name":"file-root:/etc/zipkin/ca.pem","sds_config":{"api_config_source":{"api_type":"GRPC","grpc_services":[{"envoy_grpc":{"cluster_name":"sds-grpc"}}],"transport_api_version":"V3"},"resource_api_version":"V3"}}}},"sni":"tracer"}},
+        "transport_socket": {"name":"envoy.transport_sockets.tls","typed_config":{"@type":"type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext","common_tls_context":{"combined_validation_context":{"default_validation_context":{},"validation_context_sds_secret_config":{"name":"file-root:/etc/zipkin/ca.pem","sds_config":{"api_config_source":{"api_type":"GRPC","grpc_services":[{"envoy_grpc":{"cluster_name":"sds-grpc"}}],"transport_api_version":"V3"},"initial_fetch_timeout":"0s","resource_api_version":"V3"}}}},"sni":"tracer"}},
         
         "type": "STRICT_DNS",
         "respect_dns_ttl": true,
```

`make update-golden` will generate the changes, then the other is to make the test output readable when it fails

**Comment by williamaronli** [comment]
Thanks to howard, @adiprerepa  could you do the modification following his suggestion. Thanks

**Comment by adiprerepa** [comment]
done.

**Comment by howardjohn** [comment]
@JimmyCYJ @myidpt

**Comment by williamaronli** [comment]
all looks good to me. Ask my team member to review it

File: pilot/pkg/security/model/authentication.go [hunk 2]
```diff
@@ -188,7 +180,7 @@ func ConstructSdsSecretConfig(name string) *tls.SdsSecretConfig {
 			// set the fetch timeout to 0 here because workload certs are
 			// guaranteed to exist. while others like gateway certs may not
 			// exist.
-			InitialFetchTimeout: fetchTimeout,
+			InitialFetchTimeout: ptypes.DurationProto(time.Second * 0),
```
Comment (williamaronli) [inline pilot/pkg/security/model/authentication.go:183]: why do you change the fetchTimeout here? This one should remain the old one. As you will only change the InitialFetchTimeout to 0 when the name is SDSRootResourceName or SDSDefaultResourceName

Comment (williamaronli) [reply pilot/pkg/security/model/authentication.go:183]: @adiprerepa Sorry just find that you change here could you fix this logic?

Comment (adiprerepa) [reply pilot/pkg/security/model/authentication.go:183]: oh, so only set to 0 when the name is `SDSRootResourceName` or `SDSDefaultResourceName`? got it

**Comment by adiprerepa** [comment]
@williamaronli this should do it.

File: pilot/pkg/security/model/authentication.go (hunk 2 above)
Comment (williamaronli) [reply pilot/pkg/security/model/authentication.go]: yes no need to change it here. as it is only for rootca and default case you will set the InitialFetchTimeout to 0. change it back I think will be good. and all other parts are good to me thanks

File: pkg/bootstrap/option/convert_test.go [hunk 3]
```diff
@@ -100,13 +101,14 @@ func TestTlsContextConvert(t *testing.T) {
...
+			expectTLSCtx: `{"common_tls_context":{"tls_certificate_sds_secret_configs":[{"name":"default","sds_config":{"ConfigSourceSpecifier":{"ApiConfigSource":{"api_type":2,"transport_api_version":2,"grpc_services":[{"TargetSpecifier":{"EnvoyGrpc":{"cluster_name":"sds-grpc"}}}]}},"initial_fetch_timeout":{},"resource_api_version":2}}],"ValidationContextType":{"CombinedValidationContext":{"default_validation_context":{},"validation_context_sds_secret_config":{"name":"ROOTCA","sds_config":{"ConfigSourceSpecifier":{"ApiConfigSource":{"api_type":2,"transport_api_version":2,"grpc_services":[{"TargetSpecifier":{"EnvoyGrpc":{"cluster_name":"sds-grpc"}}}]}},"initial_fetch_timeout":{},"resource_api_version":2}}}},"alpn_protocols":["istio","h2"]},"sni":"i-should-be-sni"}`,
 		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.desc, func(t *testing.T) {
 			if got := convertToJSON(tlsContextConvert(tt.tls, tt.sni, tt.meta)); !reflect.DeepEqual(tt.expectTLSCtx, got) {
+				fmt.Println(got)
```
Comment (williamaronli) [inline pkg/bootstrap/option/convert_test.go:111]: just one nit here. is the println required here?

Comment (adiprerepa) [reply pkg/bootstrap/option/convert_test.go:111]: removed.

**Comment by ericvn** [comment]
Note that release 1.4 is EOL.

**Comment by williamaronli** [comment]
@adiprerepa could you help to cherrypick this change to branch release 1.6,1.5 and 1.4?

**Comment by howardjohn** [comment]
There will never be another Istio 1.4 release. Given
https://istio.io/latest/news/support/announcing-1.5-eol/ I recommend not
backporting to 1.5

**Comment by williamaronli** [comment]
So just import to 1.6?

**Pull request merged**